### PR TITLE
config: allow hostnames for camera addresses

### PR DESCRIPTION
**Pull request by kevans91**
This offers more flexibility, as a camera system admin only
needs a DNS assignment rather than DHCP. The syntax is still
as expected:

address = "cameras.example.com:9000"

This is especially handy in more complex topologies, where one
or more specified "camera" may actually be pointed at a system
providing a tunnel through to multiple cameras; names make it
easier to identify said gateway machine.

All IPs that the given name resolves to are tried so that we
correctly handle, e.g., a name with both AAAA and A records.

Note that addresses are no longer validated up front, but instead
simply get resolved when the connection attempt is made.

Name resolution will be aborted after ~60s of failure.

Note that my camera setup is not yet complete, so this was only tested against:
- Non-existent IP
- Unresolvable hostname
- localhost w/ port closed
- localhost w/ port open

All of these behave as expected. Please excuse my rust, as I'm quite new at it.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -127,12 +129,46 @@ fn camera_loop(
...
                 return Err(cam_err.err.into());
             }
-            _ => error!(
-                "Error streaming from camera {}, will retry in {}s: {}",
-                camera_config.name,
-                current_backoff.as_secs(),
-                cam_err.err
-            ),
+            neolink::Error::AddrResolutionError => {
+                // Special handling for halting resolution after the resolve timeout period elapses.
```
Comment (thirtythreeforty) [inline src/main.rs:133]: I disagree with the strategy of "stop trying to resolve after a while."  There are many transient conditions that could cause a name resolution failure (DNS server hosed, unpropagated, etc), and we'd like to retry after those. Is there any reason this couldn't simply be removed?

In fact, I'm on the fence about the AuthFailed bailout, for similar reasons.

File: src/config.rs [hunk 2]
```diff
@@ -45,7 +44,7 @@ pub struct CameraConfig {
     pub name: String,
 
     #[serde(rename = "address")]
-    pub camera_addr: SocketAddr,
+    pub camera_addr: String,
```
Comment (thirtythreeforty) [inline src/config.rs:47]: This change removes all validation from parsing the camera address. (That is, if `SocketAddr` failed to parse, you'd get a "malformed config file" message, which I like.)  Question: Is there any validation that can be done, or is any string potentially a valid address?

File: src/bc_protocol.rs [hunk 3]
```diff
@@ -74,29 +78,48 @@ impl<'a> From<std::sync::mpsc::RecvTimeoutError> for Error {
...
+                        continue;
+                    }
+                    err => return Err(err.into()),
+                },
+            };
+
+            debug!("Success: {}", addr);
+            self.address = Some(addr);
+            self.connection = Some(conn);
+            return Ok({});
```
Comment (thirtythreeforty) [inline src/bc_protocol.rs:119]: Nitpick: The idiom is `Ok(())` ("return an `Ok` containing `()`").

File: src/bc_protocol.rs [hunk 4]
```diff
@@ -17,8 +17,9 @@ mod connection;
 mod media_packet;
 mod time;
 
-pub struct BcCamera {
-    address: SocketAddr,
+pub struct BcCamera<T: ToSocketAddrs> {
```
Comment (thirtythreeforty) [inline src/bc_protocol.rs:20]: I would prefer not to parameterize this struct. It's basically the runtime representation of the camera - and we don't have different kinds of cameras distinguished by the way they're named.

Instead, can you modify [`BcConnection::connect`](https://github.com/thirtythreeforty/neolink/blob/master/src/bc_protocol/connection.rs#L49) to accept a `ToSocketAddrs`? Then, you could pass the string straight in from the struct.

Since we create the socket with extra options, you'll still need to iterate over them in our connect_to helper. (This is [what `TcpStream::connect(ToSocketAddrs)` does](https://doc.rust-lang.org/src/std/net/tcp.rs.html#106), so that's good behavior.)

**Review by thirtythreeforty — CHANGES_REQUESTED** [review]
Thanks for the PR! Accepting hostnames is a great suggestion. See my comments if you don't mind. (Also, if you don't have time to tackle more changes, let me know and I'm happy to run with what you have.)

File: src/bc_protocol.rs (hunk 4 above)
Comment (thirtythreeforty) [reply src/bc_protocol.rs:20]: Oh, also you'll need to remove [my previous attempt](https://github.com/thirtythreeforty/neolink/blob/32400971a532509aade690b01c84c04c757dfc1c/src/bc_protocol.rs#L84) at what you're implementing. The correct behavior is indeed to iterate over the results of the name resolution.

File: src/config.rs (hunk 2 above)
Comment (kevans91) [reply src/config.rs:47]: Yeah, so the problem here is that it's actually really hard to pass an IP/name that isn't valid. e.g. observe `ping 6`, `ping 6.6`. Even worse, the POSIX-specified name resolution interface (getaddrinfo) doesn't require any particular format, it only notes that DNS *might* be the provider.

I don't have any non-FreeBSD systems around to see if it holds for GNU/Linux, but a silly example that's perfectly valid and usable here is:
```
# ping 'x!foo'
ping: Unknown host
# echo '127.0.0.1 x!foo' >> /etc/hosts
# ping 'x!foo'
PING x!foo (127.0.0.1): 56 data bytes
64 bytes from 127.0.0.1: icmp_seq=0 ttl=64 time=0.203 ms
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=117.458 ms
...
```
I had considered doing something like the following, but had somewhat reached my daily limit of Rust at the time :-) e.g. (prepare for an abomination of a mix of pseudocode and an attempt at rust):

```
if addr matches "^[-A-za-z0-9.]:[0-9]+$" || addr.parse::<Ipv4Addr>().is_ok() || addr.parse::<Ipv6Addr>().is_ok() {
    ok
}
Err(...)
```

This avoids the problem I had in my earlier prototypes where my custom validator just .to_socket_addrs() that sucker and couldn't discern potentially transient DNS errors vs. validation errors.

IMHO it's worth validating; your userbase is probably incredibly unlikely to abuse the system resolver like that, and there's a lot more value in calling it an invalid hostname for those that wouldn't. Those of us that would fiddle with it will immediately recognize the issue if the verbiage is correct, and those that wouldn't fiddle will quickly see that there's a typo.

**Comment by kevans91** [comment]
Thanks for the review! I will implement the rest of your comments within a ~day or so. I have no compelling reason to timeout after a minute, so I will remove that.

File: src/config.rs (hunk 2 above)
Comment (thirtythreeforty) [reply src/config.rs:47]: This is a really good point. Had no idea you could put line noise in the hosts file (and have it work).

Regarding errors, I think either approach is okay as long as it loudly complains about unresolvable hostnames (in clear language - even "unresolvable hostname" is a bit obtuse). Since this is a runtime problem, and given your point above, I would lean toward accepting the config file and erroring on connect.

In other words, how you have it now :)

Comment (kevans91) [reply src/config.rs:47]: I changed up the display string of the AddrResolutionError, but it might be too verbose: "Failed to translate camera address" -> I /think/ that strikes a decent balance between "what does that mean?" and "WHAT does that mean?" -- that might also be a step in the wrong direction, though. Thoughts?

File: src/bc_protocol.rs (hunk 4 above)
Comment (kevans91) [reply src/bc_protocol.rs]: I switched it up to just providing a standard new() with some sensible default values, but then it occurred to me that you probably meant to just construct a new BcConnection if BcConnection::connect() succeeds since we will destroy and reconstruct a new BcConnection on failure anyways.

File: src/config.rs (hunk 2 above)
Comment (thirtythreeforty) [reply src/config.rs:47]: Yeah, it's tricky to come up with a good error message. I think that's a decent balance too. (I don't really have a better suggestion.) Looks good!

**Review by thirtythreeforty — APPROVED** [review]
Thanks very much! This looks great.

**Comment by kevans91** [comment]
Excellent, many thanks! :-)
